### ISSUE TITLE
GCS bucket file structure

### ISSUE THREAD

**Issue by lucasrodes** [pipgs/pipgs#16]
Should we first define whether a certain file structure in the gcs bucket is required?

I'd try to keep these sort of restrictions minimum so that the library can be flexible, but I think in this case makes sense.

An example could be:

```
gcs_bucket
│
├── [package_1]
│   ├── v0.0.1
│   │   ├── [package_1]-0.0.1.tar.gz
│   ├── v0.0.2
│   │   ├── [package_1]-0.0.2.tar.gz
...
├── [package_2]
│   ├── v0.0.1
│   │   ├── [package_2]-0.0.1.tar.gz
│   ├── v0.0.2
│   │   ├── [package_2]-0.0.2.tar.gz
...
```

**Comment by vmolina** [comment]
If we want to make it general we will need to have the following:
```
gcs_bucket
|── namespace
|     |──[package-name]
|     |     |──version
|     |     |     |──package-version.tar.gz
```
Also, do we need version subspace, as this information will be available in the package name?

Also, another important question to ask if we should keep it agnostic to the bucket, that is, allowing to just input a gs://bucket/path as a dependency

**Comment by lucasrodes** [comment]
> If we want to make it general we will need to have the following:
> 
> ```
> gcs_bucket
> |── namespace
> |     |──[package-name]
> |     |     |──version
> |     |     |     |──package-version.tar.gz
> ```

Yes, IMO packages should be hosted in a certain `gs path`, so no distinction if it is a bucket root directory or a subfolder.

> Also, do we need version subspace, as this information will be available in the package name?

If all the generated files (tar.gz, whl, etc.) for each version contain a version string of known format (as in our examples), then I'd say no.

> Also, another important question to ask if we should keep it agnostic to the bucket, that is, allowing to just input a gs://bucket/path as a dependency

Do you mean, for instance, something like:

```
$ pipgs install --url gs://bucket/path package_name
```
?

If so, I'd say this should be an option. Also defining such parameters in a config file (i.e. `~/.pipgs/config.yml`)

**Comment by lucasrodes** [comment]
UPDATE on :
> Also, do we need version subspace, as this information will be available in the package name?

I think is better to have subfolders to keep all files more organized. See for instance [v1.1.5 pandas on PyPi](https://pypi.org/project/pandas/1.1.5/#files), which groups all v1.1.5 files in what it seems a version subfolder.
